pkg/apis/certificates/validation/validation.go

```diff
@@ -50,59 +52,53 @@ func ValidateCertificateRequestName(name string, prefix bool) []string {
 func ValidateCertificateSigningRequest(csr *certificates.CertificateSigningRequest) field.ErrorList {
 	isNamespaced := false
 	allErrs := apivalidation.ValidateObjectMeta(&csr.ObjectMeta, isNamespaced, ValidateCertificateRequestName, field.NewPath("metadata"))
-	err := validateCSR(csr)
 
 	specPath := field.NewPath("spec")
-
+	err := validateCSR(csr)
 	if err != nil {
 		allErrs = append(allErrs, field.Invalid(specPath.Child("request"), csr.Spec.Request, fmt.Sprintf("%v", err)))
 	}
 	if len(csr.Spec.Usages) == 0 {
 		allErrs = append(allErrs, field.Required(specPath.Child("usages"), "usages must be provided"))
 	}
-	if csr.Spec.SignerName == nil {
+	if csr.Spec.SignerName == "" {
 		allErrs = append(allErrs, field.Required(specPath.Child("signerName"), "signerName must be provided"))
```

Comment:
Defaults are applied on reads, so if an apiserver is running this version of validation, it must be running the defaulting logic, so this field will always be set.

The only case like this to handle is *client side* if we want to ensure that a newer *client* can talk to an older *apiserver* (i.e. the changes in controller-manager and kubectl). Officially I don't think this is supported, but practically it's good to do. Regardless, this shouldn't be an issue here as the apiserver is up to date.

> Since we do not allow .spec changes on update via the strategy, ValidateCertificateSigningRequestUpdate should be updated to not validate .spec at all.

```
func ValidateCertificateSigningRequestUpdate(newCSR, oldCSR *certificates.CertificateSigningRequest) field.ErrorList {
	validationErrorList := ValidateCertificateSigningRequest(newCSR)
	metaUpdateErrorList := apivalidation.ValidateObjectMetaUpdate(&newCSR.ObjectMeta, &oldCSR.ObjectMeta, field.NewPath("metadata"))
	return append(validationErrorList, metaUpdateErrorList...)
}
```

Is the current method - do you mean we should remove calling ValidateCertificateSigningRequest altogether? This isn't code added by me, and it's standard throughout all other validators so it seemed sensible to keep it there 😄

Reply:
Unless you mean in `csrStatusStrategy` we call validate update - in the case we are still fine, because when the object is decoded it'll be defaulted before the validation functions are called, meaning signerName will still be set even if a client omits it.

Reply:
> Defaults are applied on reads, so if an apiserver is running this version of validation, it must be running the defaulting logic, so this field will always be set.

For some reason I thought we only defaulted on writes, but yeah, if we default on reads then this is a non-issue.

> Is the current method - do you mean we should remove calling ValidateCertificateSigningRequest altogether? This isn't code added by me, and it's standard throughout all other validators so it seemed sensible to keep it there 😄

We do not need to change it since it causes no issue.  I was mostly saying that validating `.spec` does not make sense since we guarantee `.spec` does not change - i.e. we have asserted that it is and will always be valid on the initial `create`.